radixdlt-engine/src/main/java/com/radixdlt/atommodel/tokens/scrypt/StakingConstraintScryptV4.java

```diff
@@ -0,0 +1,220 @@
+		}
+
+		ReducerState readOwner(PreparedValidatorUpdate preparedValidatorUpdate) throws ProcedureException {
+			if (!allowDelegationFlag.getValidatorKey().equals(preparedValidatorUpdate.getValidatorKey())) {
+				throw new ProcedureException("Not matchin validator keys");
+			}
+			return new StakePrepare(
+				tokenHoldingBucket,
+				allowDelegationFlag.getValidatorKey(),
+				preparedValidatorUpdate.getOwnerAddress()::equals
+			);
+		}
+	}
+
+	private static final class StakePrepare implements ReducerState {
+		private final TokenHoldingBucket tokenHoldingBucket;
+		private final ECPublicKey validatorKey;
+		private final Predicate<REAddr> delegateAllowed;
+
+		StakePrepare(TokenHoldingBucket tokenHoldingBucket, ECPublicKey validatorKey, Predicate<REAddr> delegateAllowed) {
+			this.tokenHoldingBucket = tokenHoldingBucket;
+			this.validatorKey = validatorKey;
+			this.delegateAllowed = delegateAllowed;
+		}
+
+		ReducerState withdraw(PreparedStake preparedStake) throws ProcedureException {
+			if (preparedStake.getAmount().compareTo(ValidatorStakeData.MINIMUM_STAKE) < 0) {
+				throw new ProcedureException(
+					"Minimum amount to stake must be >= " + ValidatorStakeData.MINIMUM_STAKE
+						+ " but trying to stake " + preparedStake.getAmount()
+				);
+			}
+			if (!preparedStake.getDelegateKey().equals(validatorKey)) {
+				throw new ProcedureException("Not matching validator keys");
+			}
+
+			if (!delegateAllowed.test(preparedStake.getOwner())) {
+				throw new ProcedureException("Delegation not allowed");
+			}
+
+			return tokenHoldingBucket.withdraw(preparedStake.getResourceAddr(), preparedStake.getAmount());
+		}
+	}
+
+	private void defineStaking(Loader os) {
+		// Stake
+		os.procedure(new ReadProcedure<>(
+			TokenHoldingBucket.class, AllowDelegationFlag.class,
+			u -> new Authorization(PermissionLevel.USER, (r, c) -> { }),
+			(s, d, r) -> {
+				var nextState = (!d.allowsDelegation())
+					? new OwnerStakePrepare(s, d)
+					: new StakePrepare(s, d.getValidatorKey(), p -> true);
+				return ReducerResult.incomplete(nextState);
+			}
+		));
+		os.procedure(new ReadProcedure<>(
+			OwnerStakePrepare.class, ValidatorOwnerCopy.class,
+			u -> new Authorization(PermissionLevel.USER, (r, c) -> { }),
+			(s, d, r) -> {
+				var nextState = s.readOwner(d);
+				return ReducerResult.incomplete(nextState);
+			}
+		));
+		os.procedure(new ReadProcedure<>(
+			OwnerStakePrepare.class, PreparedValidatorUpdate.class,
+			u -> new Authorization(PermissionLevel.USER, (r, c) -> { }),
+			(s, d, r) -> {
+				var nextState = s.readOwner(d);
+				return ReducerResult.incomplete(nextState);
+			}
+		));
+		os.procedure(new UpProcedure<>(
+			StakePrepare.class, PreparedStake.class,
+			u -> new Authorization(PermissionLevel.USER, (r, c) -> { }),
+			(s, u, c, r) -> {
+				var nextState = s.withdraw(u);
+				return ReducerResult.incomplete(nextState);
+			}
+		));
```

Comment:
Now, we can have **owner-only pool** and **public pool**. I wonder if people would be interested in private pool which accepts authorised stakers only.

The would also simplify our transition flow to 
```
TokenHoldingBucket + (READ AllowDelegationFlag: a set of allowed addresses) 
    --> StakePrepare + (UP PreparedStake)
        --> END
```